include/deal.II/lac/petsc_ts.h

```diff
@@ -0,0 +1,706 @@
+    }
+
+
+    /**
+     * Pass the preconditioning matrix only.
+     * When used with 'setup_jacobian' and 'solve_for_jacobian_system', PETSc
+     * will approximate the linear system matrix-vector product using an
+     * internal matrix-free representation.
+     * When used with 'implicit_jacobian' or 'explicit_jacobian',
+     * PETSc will use the same matrix for both preconditioning and
+     * matrix-vector products.
+     */
+    void
+    reinit_matrices(PMatrixType &P)
+    {
+      this->A = nullptr;
+      this->P = &P;
+    }
+
+    /**
+     * Pass both the linear system matrix and the preconditioning matrix.
+     */
+    void
+    reinit_matrices(AMatrixType &A, PMatrixType &P)
+    {
+      this->A = &A;
+      this->P = &P;
+    }
+
+    /**
+     * Return current time
+     */
+    double
+    get_time()
+    {
+      PetscReal t;
+      AssertTS(TSGetTime(ts, &t));
+      return t;
+    }
+
+    /**
+     * Return current time step
+     */
+    double
+    get_time_step()
+    {
+      PetscReal dt;
+      AssertTS(TSGetTimeStep(ts, &dt));
+      return dt;
+    }
+
+    /**
+     * Integrate the differential-algebraic equations.
+     * This function returns the final number of computed steps.
+     */
+    unsigned int
+    solve(VectorType &solution)
+    {
+      auto _dealii_ts_ifunction_ =
+        [](TS ts, PetscReal t, Vec x, Vec xdot, Vec f, void *ctx)
+        -> PetscErrorCode {
+        TimeStepper *myctx;
+
+        PetscFunctionBeginUser;
+        (void)ctx;
+        AssertTS(TSGetApplicationContext(ts, (void *)&myctx));
+        auto user = static_cast<TimeStepper *>(myctx);
+
+        VectorType xdealii(x);
+        VectorType xdotdealii(xdot);
+        VectorType fdealii(f);
+        user->implicit_function(t, xdealii, xdotdealii, fdealii);
+        PetscFunctionReturn(0);
+      };
+
+      auto _dealii_ts_ijacobian_ = [](TS        ts,
+                                      PetscReal t,
+                                      Vec       x,
+                                      Vec       xdot,
+                                      PetscReal s,
+                                      Mat       A,
+                                      Mat       P,
+                                      void *    ctx) -> PetscErrorCode {
+        TimeStepper *myctx;
+
+        PetscFunctionBeginUser;
+        (void)ctx;
+        AssertTS(TSGetApplicationContext(ts, (void *)&myctx));
+        auto user = static_cast<TimeStepper *>(myctx);
+
+        VectorType  xdealii(x);
+        VectorType  xdotdealii(xdot);
+        AMatrixType Adealii(A);
+        PMatrixType Pdealii(P);
+        user->implicit_jacobian(t, xdealii, xdotdealii, s, Adealii, Pdealii);
+        PetscBool flg;
+        AssertTS(PetscObjectTypeCompare((PetscObject)A, MATMFFD, &flg));
+        if (flg)
+          {
+            AssertTS(MatAssemblyBegin(A, MAT_FINAL_ASSEMBLY));
+            AssertTS(MatAssemblyEnd(A, MAT_FINAL_ASSEMBLY));
+          }
```

Comment:
Can you add a comment that explains what this block does? I can't seem to figure out what the query in `PetscObjectTypeCompare` does and why the return controls whether we need to assemble the matrix. (Specifically, why the matrix!?)